src/library/features/crates/cratetreemodel.cpp

```diff
@@ -0,0 +1,61 @@
+#include <QVector>
+
+#include "library/trackcollection.h"
+#include "library/features/crates/cratetreemodel.h"
+
+#include "util/db/dbid.h"
+
+namespace {
+
+const CrateId kRootId(-1);
+
+} // anonymous namespace
+
+CrateTreeModel::CrateTreeModel(LibraryFeature* pFeature,
+                               CrateManager* pCrates)
+    : m_pFeature(pFeature),
+      m_pCrates(pCrates) {
+}
+
+void CrateTreeModel::fillTree(const QStringList& idPaths, QMap<CrateId,TreeItem*> treeCrates) {
+    CrateId parentId, childId;
+    Crate child;
+
+    // looping through the paths we are gsonna get the parent crate and the child crate of
+    // each path (2nd to last and last IDs respectivly). Then we insert the child under the
+    // parent. Since it's sorted alphabetically the parent will always exist before the child
+    for (const auto& idPath : idPaths) {
```

Comment:
Reading crates one by one in a loop from the database does not scale well. Please add a comment that this algorithm needs to be optimized.